### PR TITLE
[deployment] feat: bump bundled NCCL to >= 2.29.7 for ncclCommSuspend/Resume

### PR DESCRIPTION
**Pull request by xiefan46**
### What does this PR do?

Bumps the bundled `nvidia-nccl-cu13` version in the stable vllm and sglang Docker images from **2.28.9** (the version torch 2.11.0 cu130 pins by default) to **>= 2.29.7**, so the resulting `libnccl.so.2` exposes the `ncclCommSuspend` / `ncclCommResume` APIs. This is the prerequisite of #6408 

  > AI assistance was used to draft this PR . Submitter has reviewed every changed line and tested this feature locally. 

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

Verified locally on Mac arm64 host via Docker Desktop with `--platform linux/amd64` (QEMU emulation). The verification builds a minimal repro that mirrors the relevant prefix of
  `docker/Dockerfile.stable.vllm` (CUDA 13.0.2 base + Python 3.12 + `pip install torch==2.11.0 --index-url .../cu130`), runs the proposed `pip install --no-deps --upgrade "nvidia-nccl-cu13>=2.29.7,<3.0"`, and
   inspects the resulting `libnccl.so.2`.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

**Comment by xiefan46** [comment]
@ETOgaosion please help to review this PR, thanks!

**Comment by ETOgaosion** [comment]
Thanks for contribution. There are some other packages requires updates, I will cherry-pick your PR and use latest docker to test CI~

**Comment by ETOgaosion** [comment]
@xiefan46 Notice that after nvidia-nccl-cu13 installed, there can be version mismatch between `torch.cuda.nccl.version()` and `nvidia-nccl-cu13`, is it acceptable?

**Comment by ETOgaosion** [comment]
Sadly, I tested for you:

```
python -c "import torch;print(torch.cuda.nccl.version())"
(2, 28, 9)
root@18c61a50cac8:/# python -c "import torch;print(torch.cuda.nccl.version())"
(2, 28, 9)
root@18c61a50cac8:/# pip show nvidia-nccl-cu13
Name: nvidia-nccl-cu13
Version: 2.30.4
Summary: NVIDIA Collective Communication Library (NCCL) Runtime
Home-page: https://developer.nvidia.com/cuda-zone
Author: Nvidia CUDA Installer Team
Author-email: compute_installer@nvidia.com
License-Expression: LicenseRef-NVIDIA-Proprietary
Location: /usr/local/lib/python3.12/dist-packages
Requires: 
Required-by: torch
root@18c61a50cac8:/# find /usr -name "libnccl.so.*"
/usr/local/lib/python3.12/dist-packages/nvidia/nccl/lib/libnccl.so.2
/usr/lib/x86_64-linux-gnu/libnccl.so.2.28.3
/usr/lib/x86_64-linux-gnu/libnccl.so.2
root@18c61a50cac8:/# python
Python 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ctypes
>>> lib = ctypes.CDLL("libnccl.so.2")
>>> hasattr(lib, "ncclCommSuspend")
False
```

We may need to find proper way to use new nccl.

You can use verlai/verl:vllm020.dev2 to try~

**Comment by xiefan46** [comment]
Hi @ETOgaosion I did some investigation and found 2 problems: 
1. torch.cuda.nccl.version() always returns the nccl version in compiled time, but it won't affect the real nccl lib used in runtime
2.  ctypes.CDLL("libnccl.so.2") loads the wrong nccl lib because multiple libnccl.so.2 exist. I made some changes in another PR to make sure that it always load the right path: https://github.com/verl-project/verl/pull/6408/changes#diff-0c24e49c92be60c27a70ec2ec98cf769e3eb3ebb76924f2e154b80176d6c1dadR101-R120 

Here is the test result

> root@d37dfe860042:/# python3 -c "import torch; print(torch.__version__)"
2.11.0+cu130
root@d37dfe860042:/# find / -name "libnccl.so*" 2>/dev/null | xargs ls -la
lrwxrwxrwx 1 root root        12 Aug 30  2025 /usr/lib/x86_64-linux-gnu/libnccl.so -> libnccl.so.2
lrwxrwxrwx 1 root root        17 Aug 30  2025 /usr/lib/x86_64-linux-gnu/libnccl.so.2 -> libnccl.so.2.28.3
-rw-r--r-- 1 root root 189646136 Aug 30  2025 /usr/lib/x86_64-linux-gnu/libnccl.so.2.28.3
-rw-r--r-- 1 root root 247269392 Jun  8 14:13 /usr/local/lib/python3.12/dist-packages/nvidia/nccl/lib/libnccl.so.2
root@d37dfe860042:/# python3 -c 'import psutil, torch; print("\n".join(sorted({m.path for m in psutil.Process().memory_maps() if "libnccl" in m.path.lower()})))'
/usr/local/lib/python3.12/dist-packages/nvidia/nccl/lib/libnccl.so.2
root@d37dfe860042:/# python3 -c "
import ctypes
lib = ctypes.CDLL('/usr/local/lib/python3.12/dist-packages/nvidia/nccl/lib/libnccl.so.2')
gv = lib.ncclGetVersion; gv.argtypes=[ctypes.POINTER(ctypes.c_int)]
v = ctypes.c_int(0); gv(ctypes.byref(v))
print(f'runtime ver: {v.value//10000}.{(v.value%10000)//100}.{v.value%100}')
print(f'ncclCommSuspend: {hasattr(lib, \"ncclCommSuspend\")}')
print(f'ncclCommResume:  {hasattr(lib, \"ncclCommResume\")}')
"
runtime ver: 2.30.4
ncclCommSuspend: True
ncclCommResume:  True
root@d37dfe860042:/# python3 -c "import torch; print(torch.cuda.nccl.version())"
(2, 28, 9)

with changes in another PR, no extra change is needed in this PR. Could you help to take a look at this PR again? Thanks!

**Comment by ETOgaosion** [comment]
Let's move to https://github.com/verl-project/verl/pull/6551 , have cherry-pick your commits~

**Comment by ji-huazhong** [comment]
Closed by https://github.com/verl-project/verl/pull/6551
